### PR TITLE
Add resize support for ceph RBD

### PR DESCRIPTION
**Pull request by NickrenREN**
Add resize support for ceph RBD
**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: part of [#657](https://github.com/kubernetes/community/pull/657)

**Special notes for your reviewer**:

**Release note**:
```release-note
Add resize support for ceph RBD
```

WIP, need to add fs resize,
assign to myself first

/assign @NickrenREN

### CODE REVIEW

**Comment by NickrenREN** [comment]
I test it on my local machine, and i works
nickren@nickren-14:~/test/test/rbd$ kubectl create -f rbd-pvc.yaml 
`persistentvolumeclaim "ceph-claim" created`
nickren@nickren-14:~/test/test/rbd$ kubectl create -f rbd.yaml 
`persistentvolume "ceph-pv" created`
```
nickren@nickren-14:~/test/test/rbd$ kubectl get pvc
NAME         STATUS    VOLUME    CAPACITY   ACCESS MODES   STORAGECLASS   AGE
ceph-claim   Bound     ceph-pv   1Gi        RWO            fast           42s
```

expand pvc request:
```
nickren@nickren-14:~/test/test/rbd$ kubectl edit pvc ceph-claim
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/rbd
  creationTimestamp: 2017-10-19T06:09:32Z
  name: ceph-claim
  namespace: default
  resourceVersion: "3116"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/ceph-claim
  uid: 12d5333c-b494-11e7-8c48-080027765304
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      **storage: 2Gi**
  storageClassName: fast
  volumeName: ceph-pv
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  phase: Bound
~
```

pv capacity changed
```
nickren@nickren-14:~/test/test/rbd$ kubectl get pv
NAME      CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
ceph-pv   1Gi        RWO            Retain           Bound     default/ceph-claim   fast                     1m
nickren@nickren-14:~/test/test/rbd$ kubectl get pv
NAME      CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
ceph-pv   2Gi        RWO            Retain           Bound     default/ceph-claim   fast                     2m
nickren@nickren-14:~/test/test/rbd$ kubectl get pv
NAME      CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
ceph-pv   2Gi        RWO            Retain           Bound     default/ceph-claim   fast                     3m
```

**Comment by mingshuaiwang** [comment]
hi, any idea about fs resize?

**Comment by NickrenREN** [comment]
@mingshuaiwang  @gnufied is working on that

**Comment by gnufied** [comment]
I am not super familiar with the way we are fetching secrets etc. cc @rootfs  for taking a look.

File: pkg/volume/rbd/rbd_util.go [hunk 1]
```diff
@@ -441,6 +445,123 @@ func (util *RBDUtil) DeleteImage(p *rbdVolumeDeleter) error {
...
+		return 0, err
+	}
+
+	if len(output) == 0 {
+		return 0, fmt.Errorf("can not get image size info %s: %s", b.Image, output)
+	}
+
+	// get the size value string, just between `size ` and ` MB in`, such as `size 1024 MB in 256 objects`
+	sizeIndex := strings.Index(output, imageSizeStr)
+	divIndex := strings.Index(output, sizeDivStr)
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: use regex

File: pkg/volume/rbd/rbd_util.go [hunk 2]
```diff
@@ -441,6 +445,123 @@ func (util *RBDUtil) DeleteImage(p *rbdVolumeDeleter) error {
...
+	if err != nil {
+		return 0, err
+	}
+
+	if len(output) == 0 {
+		return 0, fmt.Errorf("can not get image size info %s: %s", b.Image, output)
+	}
+
+	// get the size value string, just between `size ` and ` MB in`, such as `size 1024 MB in 256 objects`
+	sizeIndex := strings.Index(output, imageSizeStr)
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: why not just parse json?

File: pkg/volume/rbd/rbd_util.go [hunk 3]
```diff
@@ -441,6 +445,123 @@ func (util *RBDUtil) DeleteImage(p *rbdVolumeDeleter) error {
...
+	newVolSz := fmt.Sprintf("%d", sz)
+	newSizeQuant := resource.MustParse(fmt.Sprintf("%dMi", sz))
+
+	// check the current size of rbd image, if equals to or greater that the new request size, do nothing
+	curSize, infoErr := util.rbdInfo(rbdExpander.rbdMounter)
+	if infoErr != nil {
+		return oldSize, fmt.Errorf("rbd info failed, error: %v", infoErr)
+	}
+	if curSize >= sz {
+		return newSizeQuant, nil
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: return an error instead? invalid resize

Comment (gnufied) [reply pkg/volume/rbd/rbd_util.go]: No, we want this function to be idempotent. If current disk size already satisfies user requested size then resize should be NOOP and success should be returned.

**Comment by rootfs** [comment]
prefer a better output parsing but it can come up as a followup.

let @gnufied make the final decision

File: pkg/volume/rbd/rbd_util.go (hunk 2 above)
Comment (NickrenREN) [reply pkg/volume/rbd/rbd_util.go]: Yeah, i parsed json at first and found that rbd info would return different items with different version. So i changed to cut string in the end

Comment (NickrenREN) [reply pkg/volume/rbd/rbd_util.go]: will look into this in a separate PR

**Comment by NickrenREN** [comment]
hold for retesting

**Comment by NickrenREN** [comment]
I tested it again, it works:
```
nickren@nickren-14:~/test/test/rbd$ kubectl create -f ceph-secret.yaml 
secret "ceph-secret" created
nickren@nickren-14:~/test/test/rbd$ kubectl create -f rbd-sc.yaml 
storageclass "fast" created
nickren@nickren-14:~/test/test/rbd$ kubectl create -f rbd.yaml 
persistentvolume "ceph-pv" created
nickren@nickren-14:~/test/test/rbd$ kubectl create -f rbd-pvc.yaml 
persistentvolumeclaim "ceph-claim" created
nickren@nickren-14:~/test/test/rbd$ kubectl get pvc
NAME         STATUS    VOLUME    CAPACITY   ACCESS MODES   STORAGECLASS   AGE
ceph-claim   Bound     ceph-pv   1Gi        RWO            fast           4s
```

edit pvc, and change pvc request to 2Gi
```
nickren@nickren-14:~/test/test/rbd$ kubectl edit pvc ceph-claim
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
  creationTimestamp: 2017-11-22T07:06:03Z
  name: ceph-claim
  namespace: default
  resourceVersion: "1618"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/ceph-claim
  uid: 9a2ee9b8-cf53-11e7-92d5-080027765304
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
  storageClassName: fast
  volumeName: ceph-pv
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  phase: Bound
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
~
"/tmp/kubectl-edit-fplln.yaml" 30 lines, 809 characters written
persistentvolumeclaim "ceph-claim" edited

nickren@nickren-14:~/test/test/rbd$ kubectl get pv
NAME      CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS    CLAIM                STORAGECLASS   REASON    AGE
ceph-pv   2Gi        RWO            Retain           Bound     default/ceph-claim   fast                     1m
```
PV is expanded

**Comment by NickrenREN** [comment]
just rebase @gnufied

**Comment by NickrenREN** [comment]
@enisoc could you help readd 1.9 milestone ?  It was added before and robot remove it out because it did not have urgent label

**Comment by derekwaynecarr** [comment]
I only looked at the admission change.

/approve

**Comment by left2right** [comment]
will this feature be included in release 1.9 ?

**Comment by NickrenREN** [comment]
yeah

**Comment by left2right** [comment]
thx，i want to test and use this feature in my environment, which branch should i use,release-1.9 or  master? @NickrenREN

**Comment by NickrenREN** [comment]
Both are ok. PR is merged @left2right 
Feel free to ping me If there is a  bug or sth like that

**Comment by left2right** [comment]
kubectl version
Client Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.0-alpha.3", GitCommit:"1ced91f2011b6d20a9572cf782762a1ca5da05c0", GitTreeState:"clean", BuildDate:"2017-11-15T23:06:17Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.0-alpha.3", GitCommit:"1ced91f2011b6d20a9572cf782762a1ca5da05c0", GitTreeState:"clean", BuildDate:"2017-11-15T22:53:59Z", GoVersion:"go1.9.1", Compiler:"gc", Platform:"linux/amd64"}

when use `kubectl edit pvc pika-single-pika-0` resize my pvc on ceph rbd, get this error : 
persistentvolumeclaims "pika-single-pika-0" was not valid:
# * spec: Forbidden: field is immutable after creation

@NickrenREN thx

**Comment by left2right** [comment]
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# persistentvolumeclaims "pika-single-pika-0" was not valid:
# * spec: Forbidden: field is immutable after creation
#
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/rbd
  creationTimestamp: 2017-12-26T09:20:20Z
  labels:
    app: pika
  name: pika-single-pika-0
  namespace: default
  resourceVersion: "1543731"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/pika-single-pika-0
  uid: fee94421-ea1d-11e7-8a08-005056b37ed3
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: fast-rbd
  volumeName: pvc-fee94421-ea1d-11e7-8a08-005056b37ed3
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 5Gi
  phase: Bound

**Comment by left2right** [comment]
```
# Please edit the object below. Lines beginning with a '#' will be ignored,
# and an empty file will abort the edit. If an error occurs while saving this file will be
# reopened with the relevant failures.
#
# persistentvolumeclaims "pika-single-pika-0" was not valid:
# * spec: Forbidden: field is immutable after creation
#
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    pv.kubernetes.io/bind-completed: "yes"
    pv.kubernetes.io/bound-by-controller: "yes"
    volume.beta.kubernetes.io/storage-provisioner: kubernetes.io/rbd
  creationTimestamp: 2017-12-26T09:20:20Z
  labels:
    app: pika
  name: pika-single-pika-0
  namespace: default
  resourceVersion: "1543731"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/pika-single-pika-0
  uid: fee94421-ea1d-11e7-8a08-005056b37ed3
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 10Gi
  storageClassName: fast-rbd
  volumeName: pvc-fee94421-ea1d-11e7-8a08-005056b37ed3
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 5Gi
  phase: Bound
```

**Comment by NickrenREN** [comment]
Have you enabled the feature gate `ExpandPersistentVolumes` ? @left2right

**Comment by left2right** [comment]
I thought the feature was enabled by default, so i just deployed the new version kubernetes, and didn't enable it, could you tell me how to enable it, or any docs could i refer, thx? @NickrenREN

**Comment by left2right** [comment]
kube-controller-manager:
```
/usr/bin/kube-controller-manager --logtostderr=true --v=0 --master=https://master:443 --kubeconfig=/etc/kubernetes/controller-manager.kubeconfig --service-account-private-key-file=/etc/kubernetes/certs/server.key --root-ca-file=/etc/kubernetes/certs/ca.crt --feature-gates ExpandPersistentVolumes=true
```

StorageClass:
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: fast-rbd
provisioner: kubernetes.io/rbd
parameters:
  monitors: 172.16.100.4:6789,172.16.100.5:6789,172.16.100.6:6789
  adminId: admin
  adminSecretName: ceph-secret
  adminSecretNamespace: default
  pool: rbd
  userId: admin
  userSecretName: ceph-secret
  fsType: ext4
  imageFormat: "2"
  imageFeatures: "layering"
allowVolumeExpansion: true
```

StatefulSet:
```
apiVersion: "apps/v1beta1"
kind: StatefulSet
metadata:
  name: pika
spec:
  serviceName: pika
  replicas: 1
  template:
    metadata:
      labels:
        app: pika
    spec:
      imagePullSecrets:
      - name: harbor-registry
      containers:
      - name: pika
        image: registry-center.easemob.com/easemob/pika:2.2.3
        imagePullPolicy: IfNotPresent
        ports:
        - containerPort: 9221
          name: service-port
        command: ["pika"]
        args: ["-c","conf/pika.conf"]
        resources:
          limits:
            cpu: "1"
            memory: 5Gi
          requests:
            cpu: "0.1"
            memory: 0.1Gi
        volumeMounts:
        - name: pika-single
          mountPath: /pika/output/db
  volumeClaimTemplates:
  - metadata:
      name: pika-single
    spec:
      accessModes: [ "ReadWriteOnce" ]
      storageClassName: fast-rbd
      resources:
        requests:
          storage: 5Gi
```
@NickrenREN

**Comment by NickrenREN** [comment]
Not only `controller`, but also `apiserver`, `scheduler` and `kubelet` @left2right

**Comment by left2right** [comment]
yes ,after enable kube-controller-manager, kube-scheduler, kube-apiserver, kubelet, kube-proxy' -feature-gates ExpandPersistentVolumes=true,  resize success.
however, if you use kubernetes version that doesn't have this PR commit, and use ceph rbd resize, you will get one kube-controller-manager panic error ,which @NickrenREN  fixed in this PR https://github.com/kubernetes/kubernetes/pull/57675

**Pull request merged**